source/common/http/conn_manager_impl.cc

```diff
@@ -965,9 +964,15 @@ void ConnectionManagerImpl::ActiveStream::addDecodedData(ActiveStreamDecoderFilt
     // Inline processing happens in the decodeHeaders() callback if necessary.
     filter.commonHandleBufferData(data);
   } else if (state_.filter_call_state_ & FilterCallState::DecodeTrailers) {
+    bool saved_iterate_from_current_filter = filter.iterate_from_current_filter_;
```

Comment:
nit: const

Is this actually needed now that you have the `always_start_next` param?

Reply:
Added const.

I think always_start_next is still needed in this case: a filter stops all for headers, and its decodeTrailers() calls addDecodedData(). When this happens, always_start_next is false, and iterate_from_current_filter_ is true. This is to make sure decodeTrailers() iterates from the current filter. But the decodeData() triggerred by addDecodedData() should always start from the next filter.

Reply:
Right, but doesn't decodeData() pass always_start_next to the prefix function? If so, why do we need this recursive state save?

Reply:
Ideally, we should pass always_start_next to prefix function when decodeData() is called from addDecodedData(), and pass can_start_from_current when called anywhere else. But we can't indicate it through decodeData() argument (?). So I always pass can_start_from_current to the decodeData()'s prefix function, and use iterate_from_current_filter_ to indicate which filter to start the iteration.

Another way is to remove always_start_next completely by setting iterate_from_current_filter_ after doHeader() in commonContinue() (we will need a boolean to save the state at the beginning of commonContinue()). @alyssawilk prefers to remove always_start_next too. I will make the change. But either way, I think we will need the ugly saved state.

Reply:
If it's possible, I would much prefer you add an argument to decodeData() if that would work, as I think it's much easier to reason about then this type of recursive state save logic. Is it possible to do that?

Reply:
Yes. Introduced an argument in de[en]codeData() to remove the saved state. It does look cleaner. Thanks!